### PR TITLE
WIP - adds json structure for dependency install

### PR DESCRIPTION
**Pull request by GreaterGoodest**
Not ready for pull, just looking for review of current approach. 

Thoughts on current location of tools.json?
All tools are currently in one tools.json, should they be split?
Like/dislike approach to modifying defaults/tool_paths?

### CODE REVIEW

File: pipeline/recon/config.py [hunk 1]
```diff
@@ -24,12 +25,15 @@
     "subjack": f"{Path.home()}/go/bin/subjack",
     "subjack-fingerprints": f"{Path.home()}/go/src/github.com/haccer/subjack/fingerprints.json",
     "gobuster": f"{Path.home()}/go/bin/gobuster",
-    "recursive-gobuster": f"{defaults.get('tools-dir')}/recursive-gobuster/recursive-gobuster.pyz",
+    "recursive-gobuster": f"{defaults.get('tools-dir')}/recursive-gobuster/",  # recursive-gobuster.pyz -- tool filename for ref
```
Comment (epi052) [inline pipeline/recon/config.py:28]: this change will break scanning with `--recursive`.

Most of the tool definitions are paths to the tool themselves in order to execute them as commands.  Example [here](https://github.com/epi052/recon-pipeline/blob/master/pipeline/recon/web/gobuster.py#L141)

File: pipeline/recon/tool_definitions.py [hunk 2]
```diff
@@ -1,9 +1,24 @@
+import json
 from pathlib import Path
 
 from .config import tool_paths, defaults
 
 # tool definitions for recon-pipeline's auto-installer
-tools = {
+tools = None
+with open(Path(__file__).parent / ".." / "data" / "tools.json", "r") as file:
```
Comment (epi052) [inline pipeline/recon/tool_definitions.py:8]: `Path(__file__).parent / ".."`

can be written as 

`Path(__file__).parent.parent`
or
`Path.cwd().parents[1]`

i personally prefer `Path(__file__).parent.parent` for readability, but that's 100% opinion

**Comment by GreaterGoodest** [comment]
new change structure in place and pushed.

had to use --no-verify due to some issues with flake8:
pipeline/tools/__init__.py:1:1: F401 '.loader.tools' imported but unused
pipeline/recon-pipeline.py:79:1: E402 module level import not at top of file
pipeline/recon-pipeline.py:79:1: F401 '.tools.tools' imported but unused

tests are all passing, besides skipping an nmap test. I think this is causing a change in coverage reporting:
tests/test_recon/test_nmap.py ..........s 
pipeline/recon/nmap.py                       121     20    83%   282-324

File: pipeline/recon-pipeline.py [hunk 3]
```diff
@@ -77,6 +76,8 @@ def cluge_package_imports(name, package):
     port_results_parser,
 )
 
+from .tools import tools
```
Comment (epi052) [inline pipeline/recon-pipeline.py:79]: `from .tools import tools  # noqa: F401,E402`

adding the comment at the end will suppress the flake8 warnings.  The imports are after logical code to allow for an improved command line user experience.  We do this with forethought and malice, and accept that we are breaking flake8 recommendations in doing so.

File: pipeline/recon-pipeline.py [hunk 4]
```diff
@@ -302,6 +303,7 @@ def do_install(self, args):
             # get all of the requested tools dependencies
 
             for dependency in tools.get(args.tool).get("dependencies"):
+                print(dependency)
```
Comment (epi052) [inline pipeline/recon-pipeline.py:306]: this probably doesn't need to be here 😃

File: pipeline/tools/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+from .loader import tools
```
Comment (epi052) [inline pipeline/tools/__init__.py:1]: `from .loader import tools # noqa: F401,E402`

**Comment by epi052** [comment]
Looks **Good**, thanks for this!

**Pull request merged**